clamp/clamp.tex

```diff
@@ -0,0 +1,341 @@
+      \False & \textrm{otherwise}
+   \end{cases}
+\end{equation}
+
+\section{Proof}
+\subsection{Symmetric Distance}
+%\todo{Assumptions on min and max functions? Only min is defined in Rust, and assumes PartialOrd.}
+
+%\todo{The carry type in Rust; e.g., bounded R is f32. But since it is bounded, it must be between L and U, but this does not follow from f32.} 
+
+%\todo{Change theorem statement to include all needed  so that the theorems are claims about the constructors}
+
+%\silvia{Reference specific lines in the pseudocode? E.g., for function.}
+\begin{theorem}
+    For every setting of the input parameters ($L, U$, metric) to \texttt{MakeClamp} such that the given preconditions
+    hold, the transformation returned by \texttt{MakeClamp} has the following properties:
+    \begin{enumerate}
+        \item \textup{(Appropriate output domain).} For every vector $v$ in the input domain, $\function(v)$ is in the output domain. % Prof. Vadhan said on 29/6 to leave it as this for now, but maybe in the future we add to the theorem statement what exactly the input domain and output domain are (although of course we already say this in the proof)
+        \item \textup{(Stability guarantee).} For every input $v,w$ of type \texttt{Vec(float)} and for every pair $(\din, \dout)$ (appropriately quantified), if $v,w$ are $d_{in}$-close under the symmetric distance metric and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $d_{out}$-close under the symmetric distance metric.
+    \end{enumerate}
+\end{theorem}
+
+\begin{proof}
+\textbf{(Appropriate output domain).} In the case of \texttt{MakeClamp}, this corresponds to showing that for every vector $v$ of type \texttt{Vec(T)}, $\function(v)$ is an element of \texttt{VectorDomain} \texttt{(IntervalDomain(L, U))}. For that, we need to show two things: first, that the vector entries of $\function(v)$ have type \T. % Recall that the compiler infers the float type from VectorDomain(IntervalDomain(L, U)) because L and U have type float
+Second, that they belong to the interval $[L, U]$.
+
+That $\function(v)$ has type \texttt{Vec(T)}  follows from the assumption that $v$ is in the input domain, the precondition requirement that \T has type \T, and the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(T)} and returns an element of type \texttt{Vec(T)}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises an exception for incorrect input type. Secondly, we need to show that the vector entries belong to the interval $[L, U]$. This follows from the definition of \texttt{clamp} in line \ref{line:clamp}. According to line \ref{line:clamp} in the pseudocode, there are 3 possible cases to consider:
```

Comment:
In the PDF, `\T has` looks a bit squished together as "Thas":

<img width="117" alt="Screen Shot 2021-07-08 at 12 05 01 PM" src="https://user-images.githubusercontent.com/21006/124955156-d5f3e300-dfe4-11eb-9e06-9a89b75c872c.png">

Later such squishing include "Tis" and "Tby".

Reply:
Fixed, thank you for the catch